### PR TITLE
Pino logger with Opensearch

### LINKED ISSUE #8022: Better Error Tracking Tooling
One option: https://trackjs.com/

### PR DESCRIPTION
**Pull request by ashish-aesthisia**
## Summary

Opensearch integration with pino logger for server-core by using package pino-opensearch

### To be merged with:
https://github.com/puttclub/puttclub/pull/124

## References

#8022 

## Checklist
- [ ] If this PR is still a WIP, convert to a draft
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] [ensure all checks pass](https://github.com/etherealengine/etherealengine/wiki/Testing-&-Contributing)
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewer

## QA Steps

Need to deploy opensearch node for testing in local with command
`docker run -d -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node" -e "plugins.security.disabled=true" opensearchproject/opensearch:latest`

### CODE REVIEW

**Comment by hanzlamateen** [comment]
@ashish-aesthisia so once I run container, update my .env.local file and run the engine. How can I view the logs? Do I need to download https://opensearch.org/docs/latest/install-and-configure/install-dashboards/windows/ Or is there some api call that I can use to retrieve logs?

**Comment by ashish-aesthisia** [comment]
> @ashish-aesthisia so once I run container, update my .env.local file and run the engine. How can I view the logs? Do I need to download https://opensearch.org/docs/latest/install-and-configure/install-dashboards/windows/ Or is there some api call that I can use to retrieve logs?

if you are using docker, then you can deploy opensearch on docker using this command

`docker run -d -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node" -e "plugins.security.disabled=true" opensearchproject/opensearch:latest`

After that you can check the index & log
Screenshot 2023-08-23 at 5 08 35 PM
s  using curl as given in this screenshot
Screenshot 2023-08-23 at 5 04 12 PM

**Comment by hanzlamateen** [comment]
So when it seems like client logs are not getting logged. I tried share button in share menu as it calls' code of `logger.info()`. The logger.info method gets executed but when I ran search command I was unable to see the log "successfully shared".

```
hanzlamateen@hanzla-pc:~/etherealengine$ curl -X GET "http://localhost:9200/ethereal/_search" -u "admin:admin"
{"took":2,"timed_out":false,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0},"hits":{"total":{"value":307,"relation":"eq"},"max_score":1.0,"hits":[{"_index":"ethereal","_id":"XjybQIoBbL8ZHFyRxBQP","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:23.599Z","component":"server-core:sequelize","msg":"Starting app."}},{"_index":"ethereal","_id":"XzybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:23.698Z","component":"server-core:mysql","msg":"Starting app."}},{"_index":"ethereal","_id":"YDybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.186Z","component":"server-core","msg":"Feathers-sync started."}},{"_index":"ethereal","_id":"YTybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.189Z","component":"server-core:user","msg":"Starting server with HTTPS"}},{"_index":"ethereal","_id":"YjybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.249Z","component":"server-core:user","msg":"Feathers application started on https://localhost:3030"}},{"_index":"ethereal","_id":"YzybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.253Z","component":"server-core","msg":"Running CORS on port \"3029\"."}},{"_index":"ethereal","_id":"ZDybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.417Z","component":"server-core:sequelize","msg":"Server Ready"}},{"_index":"ethereal","_id":"ZTybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.501Z","component":"server-core","msg":"597dd0aaf6be9b7675a437c9ee644f71a2eef8ef\n"}},{"_index":"ethereal","_id":"ZjybQIoBbL8ZHFyRxBQZ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.545Z","component":"server-core","msg":"òòòòòò 597dd0aaf6be9b7675a437c9ee644f71a2eef8ef ò 2023-08-24T06:04:02+00:00 ò type fix ò HEAD -> 8022-opensearch-integration-for-logging, origin/8022-opensearch-integration-for-logging ò  ò ashish@aesthisia.com ò ashish@aesthisia.com òò"}},{"_index":"ethereal","_id":"ZzybQIoBbL8ZHFyRxBQa","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.566Z","component":"server-core","msg":"uploadLocalProjectToProvider for project \"default-project\" started at \"Tue Aug 29 2023 14:24:24 GMT+0500 (Pakistan Standard Time)\"."}}]}}
```

image

**Comment by ashish-aesthisia** [comment]
> So when it seems like client logs are not getting logged. I tried share button in share menu as it calls' code of `logger.info()`. The logger.info method gets executed but when I ran search command I was unable to see the log "successfully shared".
> 
> ```
> hanzlamateen@hanzla-pc:~/etherealengine$ curl -X GET "http://localhost:9200/_cat/indices?v" -u "admin:admin"
> health status index                     uuid                   pri rep docs.count docs.deleted store.size pri.store.size
> yellow open   .plugins-ml-config        QI0XdjMbTZOJrA4EUo_6_A   1   1          1            0      3.9kb          3.9kb
> green  open   .opensearch-observability KNCxxhPHQ1KDtzGtpENqdw   1   0          0            0       208b           208b
> yellow open   ethereal                  07-Z_W7cQUiRRtWoOADyTw   1   1        307            0      117kb          117kb
> hanzlamateen@hanzla-pc:~/etherealengine$ curl -X GET "http://localhost:9200/ethereal/_search" -u "admin:admin"
> {"took":2,"timed_out":false,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0},"hits":{"total":{"value":307,"relation":"eq"},"max_score":1.0,"hits":[{"_index":"ethereal","_id":"XjybQIoBbL8ZHFyRxBQP","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:23.599Z","component":"server-core:sequelize","msg":"Starting app."}},{"_index":"ethereal","_id":"XzybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:23.698Z","component":"server-core:mysql","msg":"Starting app."}},{"_index":"ethereal","_id":"YDybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.186Z","component":"server-core","msg":"Feathers-sync started."}},{"_index":"ethereal","_id":"YTybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.189Z","component":"server-core:user","msg":"Starting server with HTTPS"}},{"_index":"ethereal","_id":"YjybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.249Z","component":"server-core:user","msg":"Feathers application started on https://localhost:3030"}},{"_index":"ethereal","_id":"YzybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.253Z","component":"server-core","msg":"Running CORS on port \"3029\"."}},{"_index":"ethereal","_id":"ZDybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.417Z","component":"server-core:sequelize","msg":"Server Ready"}},{"_index":"ethereal","_id":"ZTybQIoBbL8ZHFyRxBQQ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.501Z","component":"server-core","msg":"597dd0aaf6be9b7675a437c9ee644f71a2eef8ef\n"}},{"_index":"ethereal","_id":"ZjybQIoBbL8ZHFyRxBQZ","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.545Z","component":"server-core","msg":"òòòòòò 597dd0aaf6be9b7675a437c9ee644f71a2eef8ef ò 2023-08-24T06:04:02+00:00 ò type fix ò HEAD -> 8022-opensearch-integration-for-logging, origin/8022-opensearch-integration-for-logging ò  ò ashish@aesthisia.com ò ashish@aesthisia.com òò"}},{"_index":"ethereal","_id":"ZzybQIoBbL8ZHFyRxBQa","_score":1.0,"_source":{"level":30,"time":"2023-08-29T09:24:24.566Z","component":"server-core","msg":"uploadLocalProjectToProvider for project \"default-project\" started at \"Tue Aug 29 2023 14:24:24 GMT+0500 (Pakistan Standard Time)\"."}}]}}
> ```
> 
> image

Who can confirm this? if client logs are being sent to server logger? Last time when I & josh checked, we found out the it happens with the logging api.

**Comment by hanzlamateen** [comment]
> 

We can look into this tomorrow in voice chat

**Review by HexaField — CHANGES_REQUESTED** [review]

File: .env.local.default [hunk 1]
```diff
@@ -24,7 +24,7 @@ APP_PORT=3000
 VITE_HMR=false
 
 # Vite variables - Exposed to client side
-VITE_FORCE_CLIENT_LOG_AGGREGATE=false
+VITE_FORCE_CLIENT_LOG_AGGREGATE=true
```
[the remark this answers is not in the document]
Comment (hanzlamateen) [reply .env.local.default:27]: Agreed

**Comment by hanzlamateen** [comment]
@ashish-aesthisia this branch has conflicts with dev. Can you fix them?

File: packages/engine/src/common/functions/logger.ts [hunk 2]
```diff
@@ -29,21 +29,29 @@ Ethereal Engine. All Rights Reserved.
...
 
-import fetch from 'cross-fetch'
-import LruCache from 'lru-cache'
+import { ServiceTypes } from '@etherealengine/common/declarations'
+import config from '@etherealengine/common/src/config'
+import { FeathersApplication } from '@feathersjs/feathers'
+import { logsApiPath } from '../../schemas/cluster/logs-api.schema'
 
-import config from './config'
+// const logRequestCache = new LruCache({
```
Comment (hanzlamateen) [inline packages/engine/src/common/functions/logger.ts:43]: @HexaField so we have fixed client logs being sent to server and opensearch. But by looking into the number of requests being sent from client to `logs-api` service, I think we should have some sort of caching on client end.

**Review by hanzlamateen — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -105,6 +105,7 @@
     "@swc/core": "1.3.41",
     "@types/app-root-path": "1.2.4",
     "@types/cli": "0.11.21",
+    "@types/lru-cache": "^7.10.10",
```
Comment (hanzlamateen) [inline package.json:108]: I think this and `"lru-cache"` should be in engine's `package.json`

File: packages/engine/src/common/functions/logger.ts [hunk 4]
```diff
@@ -95,36 +105,47 @@ const multiLogger = {
...
+
+            // In addition to sending to logging endpoint,  output to console
+            consoleMethods[level](...args)
+
+            // Send an async rate-limited request to backend logs-api service for aggregation
+            // Also suppress logger.info() levels (the equivalent to console.log())
+
+            if (config.client.serverHost && level !== 'info') {
+              logRequestCache.set(logParams.msg, () =>
+                fetch(url, {
```
Comment (hanzlamateen) [inline packages/engine/src/common/functions/logger.ts:129]: `url` points to `'/api/log'` which is no longer there. So I think this caching is not working.

File: packages/engine/src/common/functions/logger.ts [hunk 5]
```diff
@@ -95,36 +105,47 @@ const multiLogger = {
       // For non-local builds, this send() is used
       const send = (level) => {
         const url = new URL('/api/log', config.client.serverUrl)
```
Comment (hanzlamateen) [inline packages/engine/src/common/functions/logger.ts:107]: We can get rid of this `const url ...` line as its no longer required.

**Review by hanzlamateen — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (ashish-aesthisia) [reply package.json:108]: Done

File: packages/engine/src/common/functions/logger.ts (hunk 4 above)
Comment (ashish-aesthisia) [reply packages/engine/src/common/functions/logger.ts]: Done

**Review by HexaField — APPROVED** [review]

**Comment by HexaField** [comment]
Looks good!

**Review by hanzlamateen — CHANGES_REQUESTED** [review]
Does not seems to be sending any client logs:

https://github.com/EtherealEngine/etherealengine/assets/10975502/68262233-b625-4c69-a1e0-b92f31a868f8

**Comment by ashish-aesthisia** [comment]
> Does not seems to be sending any client logs:
> 
>  Recording.2023-09-18.204534.mp4
Screenshot 2023-09-19 at 11 02 55 AM

Recentely checked getting the client logs too.

**Review by hanzlamateen — APPROVED** [review]
Looks good. At some point we should add an unload/close event handler to send client logs to the server if the browser window/tab was closed.

**Pull request merged**